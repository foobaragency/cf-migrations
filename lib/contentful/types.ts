export type LocaleDependent = { [locale: string]: string }

export type MigrationEntry = {
  fileName: LocaleDependent
}
