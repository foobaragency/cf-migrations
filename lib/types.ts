import Migration from "contentful-migration"

export { Migration }

export type ContentfulPartialOptions = {
  environmentId: string
  spaceId: string
  accessToken: string
  locale?: string
}

export type MigrationOptions = ContentfulPartialOptions & {
  migrationsDirectory: string
}
