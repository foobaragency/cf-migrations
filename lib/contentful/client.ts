import { createClient, PlainClientAPI } from "contentful-management"

import { ContentfulPartialOptions } from "../types"

let client: PlainClientAPI | undefined

export function getClient({ accessToken }: ContentfulPartialOptions) {
  if (!client) {
    client = createClient({ accessToken }, { type: "plain" })
  }

  return client
}
