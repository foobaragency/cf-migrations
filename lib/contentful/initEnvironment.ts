import path from "path"

import { runMigration } from "contentful-migration"

import { ContentfulPartialOptions } from "../types"

export async function initEnvironment(options: ContentfulPartialOptions) {
  const filePath = path.normalize(
    `${process.cwd()}/node_modules/cf-migrations/dist/migrations/0001-create-migration-type.js`
  )
  await runMigration({
    ...options,
    filePath,
    yes: true,
  })
}
