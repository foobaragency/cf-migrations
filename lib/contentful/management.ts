import {
  CollectionProp,
  EntryProps,
} from "contentful-management/dist/typings/export-types"

import { config } from "../config"
import { ContentfulPartialOptions } from "../types"

import {
  createMigrationEntries,
  getContentTypes,
  getMigrationEntries,
} from "./migrationEntries"
import { MigrationEntry } from "./types"

export async function updateMigrationState(
  options: ContentfulPartialOptions,
  migrationStates: MigrationEntry[]
) {
  for (const stateFields of migrationStates) {
    await createMigrationEntries(stateFields, options)
  }
}

export async function getDeployedMigrations(options: ContentfulPartialOptions) {
  const migrationEntries = await getMigrationEntries(options)

  return getDeployedMigrationNames(migrationEntries, options.locale)
}

export function getDeployedMigrationNames(
  migrationEntries: CollectionProp<EntryProps>,
  contentfulLocale?: string
) {
  const locale = contentfulLocale || config.contentful.defaultLocale
  const deployedFiles = migrationEntries.items.map(
    item => (item.fields as MigrationEntry).fileName[locale]
  )

  return deployedFiles
}

export async function assessMigrationsTypeExistence(
  options: ContentfulPartialOptions
) {
  const contentTypes = await getContentTypes(options)

  return contentTypes.items.some(
    contentType => contentType.name === config.contentful.contentType.name
  )
}
