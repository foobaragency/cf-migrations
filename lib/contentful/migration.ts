import { runMigration } from "contentful-migration"

import { info, success } from "../logger"
import { MigrationOptions, PendingMigration } from "../types"

// eslint-disable-next-line @typescript-eslint/member-delimiter-style
export type MigrationResult = {
  successful: boolean
  fileName: string
}

export async function runMigrations(
  pendingMigrations: PendingMigration[],
  options: MigrationOptions
): Promise<MigrationResult[]> {
  const yes = options.yes !== undefined ? options.yes : true
  const migrationResult: MigrationResult[] = []
  for (const { filePath, fileName } of pendingMigrations) {
    info(`Deploying migration ${filePath}...`)
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const result = await runMigration({ ...options, filePath, yes })
    if (result) {
      success(`${filePath} migration deployed`)
      migrationResult.push({ successful: true, fileName })
    }
    migrationResult.push({ successful: false, fileName })
  }

  return migrationResult
}
