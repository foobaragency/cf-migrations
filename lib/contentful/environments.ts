import { info } from "../logger"
import { ContentfulPartialOptions } from "../types"

import { getClient } from "./client"

export async function getAllEnvironments(options: ContentfulPartialOptions) {
  const result = await getClient(options).environment.getMany(options)

  return result.items
}

export async function createEnvironment(options: ContentfulPartialOptions) {
  return getClient(options).environment.createWithId(options, {
    name: options.environmentId,
  })
}

export async function deleteEnvironment(options: ContentfulPartialOptions) {
  return getClient(options).environment.delete(options)
}

export async function getAllEnvironmentAliases(
  options: ContentfulPartialOptions
) {
  return getClient(options).environmentAlias.getMany(options)
}

export async function findEnvironment(options: ContentfulPartialOptions) {
  return getClient(options).environment.get(options)
}

export async function updateEnvironmentAlias(
  aliasEnvironmentId: string,
  targetEnvironmentId: string,
  options: ContentfulPartialOptions
) {
  const releases = await getAllEnvironmentAliases(options)
  const currentAliasVersion =
    releases.items[releases.items.length - 1].sys.version
  const { spaceId } = options

  return getClient(options).environmentAlias.update(
    {
      spaceId,
      environmentAliasId: aliasEnvironmentId,
    },
    // This property sys makes no sense in the update payload ðŸ¤”
    // But is required by the typing since some properties (e.g. `version`) are used.
    //
    // According to the API documentation, it was intended to be present only in the response.
    // ref: https://www.contentful.com/developers/docs/references/content-management-api/#/reference/environment-aliases/environment-alias/create-update-an-environment-alias/console
    //
    // So, don't mind the createdAt and updatedAt having empty strings.
    {
      environment: {
        sys: {
          type: "Link",
          linkType: "Environment",
          id: targetEnvironmentId,
        },
      },
      sys: {
        type: "Environment Alias",
        id: aliasEnvironmentId,
        version: currentAliasVersion,
        createdAt: "",
        updatedAt: "",
        space: {
          sys: {
            type: "Link",
            linkType: "Space",
            id: spaceId,
          },
        },
      },
    }
  )
}

export async function checkEnvironmentReadyStatus(
  options: ContentfulPartialOptions,
  environmentCreationSecondsTimeout: number
): Promise<boolean> {
  const environment = await findEnvironment(options)
  const isEnvironmentReady = environment.sys.status.sys.id === "ready"

  if (isEnvironmentReady) {
    return true
  }

  if (environmentCreationSecondsTimeout === 0) {
    throw new Error(
      `Unable to make sure that environment "${options.environmentId}" status is ready`
    )
  }

  info(
    `Environment ${options.environmentId} isn't read yet. Remaining seconds ${environmentCreationSecondsTimeout}...`
  )
  await delay(1000)
  const remainingSeconds = environmentCreationSecondsTimeout - 1

  return checkEnvironmentReadyStatus(options, remainingSeconds)
}

function delay(milliseconds: number) {
  return new Promise(resolve => setTimeout(resolve, milliseconds))
}
