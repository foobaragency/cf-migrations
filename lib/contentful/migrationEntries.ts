import { EntryProps } from "contentful-management/types"

import { config } from "../config"
import { ContentfulPartialOptions } from "../types"

import { getClient } from "./client"
import { MigrationEntry } from "./types"

export async function createMigrationEntries(
  fields: MigrationEntry,
  options: ContentfulPartialOptions
) {
  return getClient(options).entry.create(
    { ...options, contentTypeId: config.contentful.contentType.id },
    { fields }
  )
}

export async function getMigrationEntries(options: ContentfulPartialOptions) {
  const entries: EntryProps[] = []
  const client = getClient(options)
  let totalEntries = 0

  do {
    const response = await client.entry.getMany({
      ...options,
      query: {
        content_type: config.contentful.contentType.id,
        order: "sys.createdAt",
        limit: 100,
        skip: entries.length,
      },
    })
    entries.push(...response.items)
    totalEntries = response.total
  } while (entries.length < totalEntries)

  return entries
}

export async function getContentTypes(options: ContentfulPartialOptions) {
  return getClient(options).contentType.getMany(options)
}
