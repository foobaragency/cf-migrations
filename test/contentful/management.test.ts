import { EntryProps } from "contentful-management/types"

import { getDeployedMigrationNames } from "../../lib/contentful/management"

describe("Contentful Management", () => {
  const sys = {
    space: {
      sys: {
        type: "Link",
        linkType: "Space",
        id: "rklm4ep5nhjl",
      },
    },
    id: "ZR86DUHQcB7GQNGB9AIjt",
    type: "Asset",
    createdAt: "2020-10-26T11:17:35.480Z",
    updatedAt: "2020-10-26T11:17:35.480Z",
    environment: {
      sys: {
        id: "dev",
        type: "Link",
        linkType: "Environment",
      },
    },
    version: 1,
    contentType: { sys: { type: "", linkType: "", id: "" } },
  }

  const migrationEntries: EntryProps[] = [
    {
      fields: { fileName: { "en-US": "0004-migration.ts" } },
      sys,
    },
    {
      fields: { fileName: { "en-US": "0003-migration.ts" } },
      sys,
    },
    {
      fields: { fileName: { "en-US": "0002-migration.ts" } },
      sys,
    },
    {
      fields: { fileName: { "en-US": "0001-migration.ts" } },
      sys,
    },
  ]

  it("extracts deployed migration names from the Migration entries", () => {
    const deployedMigrations = getDeployedMigrationNames(migrationEntries)

    expect(deployedMigrations).toMatchSnapshot()
  })
})
