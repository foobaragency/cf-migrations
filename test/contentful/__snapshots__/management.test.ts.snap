// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contentful Management extracts deployed migration names from the Migration entries 1`] = `
Array [
  "0004-migration.ts",
  "0003-migration.ts",
  "0002-migration.ts",
  "0001-migration.ts",
]
`;
